File: java-client/src/main/java/com/couchbase/client/java/view/SpatialViewOptions.java
/*
 * Copyright (c) 2018 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.view;

import com.couchbase.client.core.annotation.Stability;
import com.couchbase.client.core.util.UrlQueryStringBuilder;
import com.couchbase.client.java.CommonOptions;
import com.couchbase.client.java.json.JsonArray;

public class SpatialViewOptions extends CommonOptions<SpatialViewOptions> {
  public static SpatialViewOptions DEFAULT = new SpatialViewOptions();

  /**
   * Contains all stored params.
   */
  private final UrlQueryStringBuilder params = UrlQueryStringBuilder.createForUrlSafeNames();
  private boolean development;

  public static SpatialViewOptions spatialViewOptions() {
    return new SpatialViewOptions();
  }

  private SpatialViewOptions() {
  }

  public SpatialViewOptions development(boolean development) {
    this.development = development;
    return this;
  }


  /**
   * Limit the number of the returned documents to the specified number.
   *
   * @param limit the number of documents to return.
   * @return the {@link SpatialViewOptions} object for proper chaining.
   */
  public SpatialViewOptions limit(final int limit) {
    if (limit < 0) {
      throw new IllegalArgumentException("Limit must be >= 0.");
    }
    params.set("limit", limit);
    return this;
  }

  /**
   * Skip this number of records before starting to return the results.
   *
   * @param skip The number of records to skip.
   * @return the {@link SpatialViewOptions} object for proper chaining.
   */
  public SpatialViewOptions skip(final int skip) {
    if (skip < 0) {
      throw new IllegalArgumentException("Skip must be >= 0.");
    }
    params.set("skip", skip);
    return this;
  }

  /**
   * Allow the results from a stale view to be used.
   * <p>
   * See the "Stale" enum for more information on the possible options. The
   * default setting is "update_after"!
   *
   * @param stale Which stale mode should be used.
   * @return the {@link SpatialViewOptions} object for proper chaining.
   */
  public SpatialViewOptions stale(final Stale stale) {
    params.set("stale", stale.identifier());
    return this;
  }

  /**
   * Enabled debugging on view queries.
   *
   * @return the {@link SpatialViewOptions} object for proper chaining.
   */
  public SpatialViewOptions debug(boolean debug) {
    params.set("debug", debug);
    return this;
  }

  public SpatialViewOptions startRange(final JsonArray startRange) {
    params.set("start_range", startRange.toString());
    return this;
  }

  public SpatialViewOptions endRange(final JsonArray endRange) {
    params.set("end_range", endRange.toString());
    return this;
  }

  public SpatialViewOptions range(final JsonArray startRange, final JsonArray endRange) {
    startRange(startRange);
    endRange(endRange);
    return this;
  }

  /**
   * Sets the response in the event of an error.
   * <p>
   * See the "OnError" enum for more details on the available options.
   *
   * @param onError The appropriate error handling type.
   * @return the {@link SpatialViewOptions} object for proper chaining.
   */
  public SpatialViewOptions onError(final OnError onError) {
    params.set("on_error", onError.identifier());
    return this;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("ViewQuery{");
    sb.append("params=\"").append(export()).append('"');
    if (development) {
      sb.append(", dev");
    }
    sb.append('}');
    return sb.toString();
  }


  String export() {
    return params.build();
  }

  @Stability.Internal
  public SpatialViewOptions.BuiltSpatialViewOptions build() {
    return new BuiltSpatialViewOptions();
  }

  public class BuiltSpatialViewOptions extends BuiltCommonOptions {

    public boolean development() {
      return development;
    }

    public String query() {
      return export();
    }

  }

}

File: scala-client/src/main/scala/com/couchbase/client/scala/view/SpatialViewOptions.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.scala.view

import com.couchbase.client.core.retry.RetryStrategy
import com.couchbase.client.scala.json.{JsonArray, JsonArraySafe, JsonObject}
import com.couchbase.client.scala.query.QueryOptions

import scala.concurrent.duration.Duration



/** Customize the execution of a spatial view.
  *
  * @author Graham Pople
  * @since 1.0.0
  */
case class SpatialViewOptions(
                               private[scala] val development: Option[Boolean] = None,
                               private[scala] val limit: Option[Int] = None,
                               private[scala] val skip: Option[Int] = None,
                               private[scala] val stale: Option[Stale] = None,
                               private[scala] val onError: Option[OnError] = None,
                               private[scala] val debug: Option[Boolean] = None,
                               private[scala] val startRange: Option[String] = None,
                               private[scala] val endRange: Option[String] = None,
                               private[scala] val retryStrategy: Option[RetryStrategy] = None,
                               private[scala] val timeout: Option[Duration] = None
                             ) {

  def development(value: Boolean): SpatialViewOptions = {
    copy(development = Some(value))
  }

  /**
    * Limit the number of the returned documents to the specified number.
    *
    * @return this for further chaining
    */
  def limit(value: Int): SpatialViewOptions = {
    copy(limit = Some(value))
  }

  /** Skip this number of records before starting to return the results.
    *
    * @return this for further chaining
    */
  def skip(value: Int): SpatialViewOptions = {
    copy(skip = Some(value))
  }

  /** Allow the results from a stale view to be used.
    *
    * The default is Stale.UpdateAfter.
    *
    * @return this for further chaining
    */
  def stale(value: Stale): SpatialViewOptions = {
    copy(stale = Some(value))
  }

  /** Sets the response in the event of an error.
    *
    * @return this for further chaining
    */
  def onError(value: OnError): SpatialViewOptions = {
    copy(onError = Some(value))
  }

  /** Enable debugging on view queries.
    *
    * @return this for further chaining
    */
  def debug(value: Boolean): SpatialViewOptions = {
    copy(debug = Some(value))
  }

  def startRange(value: Any): SpatialViewOptions = {
    value match {
      case v: String => copy(startRange = Some(v))
      case JsonArray | JsonArraySafe => copy(startRange = Some(value.toString))
      case _ =>
        // This will trigger a validation failure later
        copy(startRange = null)
    }
  }

  def endRange(value: Any): SpatialViewOptions = {
    value match {
      case v: String => copy(endRange = Some(v))
      case JsonArray | JsonArraySafe => copy(endRange = Some(value.toString))
      case _ =>
        // This will trigger a validation failure later
        copy(endRange = null)
    }
  }

  def range(startRange: Any, endRange: Any): SpatialViewOptions = {
    val next: SpatialViewOptions = this.startRange(startRange)
    next.endRange(endRange)
  }

  def timeout(timeout: Duration): SpatialViewOptions = {
    copy(timeout = Option(timeout))
  }

  /** Sets what retry strategy to use if the operation fails.  See [[RetryStrategy]] for details.
    *
    * @param strategy the retry strategy to use
    *
    * @return a copy of this with the change applied, for chaining.
    */
  def retryStrategy(strategy: RetryStrategy): SpatialViewOptions = {
    copy(retryStrategy = Some(strategy))
  }


  private[scala] def durationToN1qlFormat(duration: Duration) = {
    if (duration.toSeconds > 0) duration.toSeconds + "s"
    else duration.toNanos + "ns"
  }

  private[scala] def encode() = {
    val sb = new StringBuilder()

    limit.foreach(v => {
      sb.append("limit=")
      sb.append(v.toString)
      sb.append('&')
    })

    skip.foreach(v => {
      sb.append("skip=")
      sb.append(v.toString)
      sb.append('&')
    })

    stale.foreach(v => {
      sb.append("stale=")
      sb.append(v.encode)
      sb.append('&')
    })

    onError.foreach(v => {
      sb.append("on_error=")
      sb.append(v.encode)
      sb.append('&')
    })

    debug.foreach(v => {
      sb.append("debug=")
      sb.append(v.toString)
      sb.append('&')
    })

    startRange.foreach(v => {
      sb.append("start_range=")
      sb.append(v.toString)
      sb.append('&')
    })

    endRange.foreach(v => {
      sb.append("end_range=")
      sb.append(v.toString)
      sb.append('&')
    })

    sb.toString.stripSuffix("&")
  }
}




